app/controllers/lti13_controller.php

```diff
@@ -0,0 +1,149 @@
+<?php
+App::import('Lib', 'Lti13Bootstrap');
+App::import('Model', 'Lti13');
+
+use IMSGlobal\LTI\LTI_Exception;
+use IMSGlobal\LTI\LTI_OIDC_Login;
+use IMSGlobal\LTI\OIDC_Exception;
+
+/**
+ * LTI 1.3 Controller
+ *
+ * @uses      AppController
+ * @package   CTLT.iPeer
+ * @since     3.4.5
+ * @author    Steven Marshall <steven.marshall@ubc.ca>
+ * @copyright 2019 All rights reserved.
+ * @license   MIT {@link http://www.opensource.org/licenses/MIT}
+ * @link      https://www.imsglobal.org/spec/security/v1p0/#fig_oidcflow
+ */
+class Lti13Controller extends AppController
+{
+    public $uses = array('Lti13');
+
+    public function __construct()
+    {
+        parent::__construct();
+    }
+
+    public function beforeFilter()
+    {
+        $this->Auth->allow();
+    }
+
+    /**
+     * OIDC login action called by platform.
+     */
+    public function login()
+    {
+        try {
+
+            $login = LTI_OIDC_Login::new($this->Lti13->db);
+            $url = Router::url('/lti13/launch', true);
+            $redirect = $login->do_oidc_login_redirect($url);
+            $redirect->do_redirect();
+
+        } catch (OIDC_Exception $e) {
+
+            $this->Session->setFlash(sprintf("Error doing OIDC login: %s", $e->getMessage()));
+            $this->redirect(array('controller'=>'home', 'action'=>'index'));
+
+        }
+    }
+
+    /**
+     * Launch action called by platform.
+     */
+    public function launch()
+    {
+        try {
+
+            $launch = $this->Lti13->launch();
+            $data = $this->Lti13->getData($launch->get_launch_id());
+            $this->Lti13->resetLogs();
+            $this->log(json_encode($data, 448), 'lti13/launch');
+
```

Comment:
A few issues at this point.

- I believe that the launch isn't actually logging in the user like you would expect (I keep on hitting the `$this->redirect('/');` for not being logged in which is weird)
- Users are not automatically created on launch if they don't exist
- Courses are not automatically created on launch if they don't exist

Reply:
First bullet: The "LTI 1.3 launch success" message will appear after the LTI launch but it's not a user login, e.g. the user login is separate from the LTI launch.

Reply:
Second bullet: LTI launch is separate from any subsequent action. "Update roster" would modify the dB and update users from Canvas to iPeer.

Third bullet: "Update roster" would modify the dB and update courses from Canvas to iPeer.